arangod/Aql/TraversalNode.cpp

```diff
@@ -379,7 +386,84 @@ void TraversalNode::toVelocyPackHelper(VPackBuilder& nodes, unsigned flags) cons
 /// @brief creates corresponding ExecutionBlock
 std::unique_ptr<ExecutionBlock> TraversalNode::createBlock(
     ExecutionEngine& engine, std::unordered_map<ExecutionNode*, ExecutionBlock*> const&) const {
-  return std::make_unique<TraversalBlock>(&engine, this);
+  ExecutionNode const* previousNode = getFirstDependency();
+  TRI_ASSERT(previousNode != nullptr);
+  auto inputRegisters = std::make_shared<std::unordered_set<RegisterId>>();
+  auto& varInfo = getRegisterPlan()->varInfo;
+  RegisterId inputRegister = MaxRegisterId;
+  if (usesInVariable()) {
+    auto it = varInfo.find(inVariable()->id);
+    TRI_ASSERT(it != varInfo.end());
+    inputRegisters->emplace(it->second.registerId);
+    inputRegister = it->second.registerId;
+    TRI_ASSERT(getStartVertex().empty());
+  }
+  auto outputRegisters = std::make_shared<std::unordered_set<RegisterId>>();
+  std::unordered_map<OutputName, RegisterId> outputRegisterMapping;
+
+  if (usesVertexOutVariable()) {
+    auto it = varInfo.find(vertexOutVariable()->id);
+    TRI_ASSERT(it != varInfo.end());
+    TRI_ASSERT(it->second.registerId < ExecutionNode::MaxRegisterId);
+    outputRegisters->emplace(it->second.registerId);
+    outputRegisterMapping.emplace(OutputName::VERTEX, it->second.registerId);
+  }
+  if (usesEdgeOutVariable()) {
+    auto it = varInfo.find(edgeOutVariable()->id);
+    TRI_ASSERT(it != varInfo.end());
+    TRI_ASSERT(it->second.registerId < ExecutionNode::MaxRegisterId);
+    outputRegisters->emplace(it->second.registerId);
+    outputRegisterMapping.emplace(OutputName::EDGE, it->second.registerId);
+  }
+  if (usesPathOutVariable()) {
+    auto it = varInfo.find(pathOutVariable()->id);
+    TRI_ASSERT(it != varInfo.end());
+    TRI_ASSERT(it->second.registerId < ExecutionNode::MaxRegisterId);
+    outputRegisters->emplace(it->second.registerId);
+    outputRegisterMapping.emplace(OutputName::PATH, it->second.registerId);
+  }
```

Comment:
Oh, and I feel obligated to add that these three options (whether there is a vertex/edge/path out variable) *could* be made template parameters. :stuck_out_tongue_winking_eye:

Reply:
i totally agree here ;)
On my todo list.